File: android/src/main/java/com/getcapacitor/community/admob/nativeads/NativeAdFeedSessions.java
package com.getcapacitor.community.admob.nativeads;

import java.util.HashMap;
import java.util.Map;

final class NativeAdFeedSessions {

    private static final int MAX_FEEDS = 2;

    private final Map<String, String> sessions = new HashMap<>();
    private final Map<String, Long> placementSequences = new HashMap<>();

    boolean start(String feedId, String sessionId) {
        if (!sessions.containsKey(feedId) && sessions.size() >= MAX_FEEDS) return false;
        sessions.put(feedId, sessionId);
        placementSequences.put(feedId, -1L);
        return true;
    }

    boolean isCurrent(String feedId, String sessionId) {
        return sessionId.equals(sessions.get(feedId));
    }

    boolean accepts(String feedId, long sequence) {
        if (sequence <= placementSequences.getOrDefault(feedId, -1L)) return false;
        placementSequences.put(feedId, sequence);
        return true;
    }

    void remove(String feedId) {
        sessions.remove(feedId);
        placementSequences.remove(feedId);
    }

    void clear() {
        sessions.clear();
        placementSequences.clear();
    }
}

File: android/src/main/java/com/getcapacitor/community/admob/nativeads/NativeAdFeedSessions.kt
package com.getcapacitor.community.admob.nativeads

internal class NativeAdFeedSessions {
    private val sessions = mutableMapOf<String, String>()
    private val placementSequences = mutableMapOf<String, Long>()

    fun start(feedId: String, sessionId: String): Boolean {
        if (feedId !in sessions && sessions.size >= MAX_FEEDS) return false
        sessions[feedId] = sessionId
        placementSequences[feedId] = -1
        return true
    }

    fun isCurrent(feedId: String, sessionId: String): Boolean = sessions[feedId] == sessionId

    fun accepts(feedId: String, sequence: Long): Boolean {
        if (sequence <= placementSequences.getOrDefault(feedId, -1)) return false
        placementSequences[feedId] = sequence
        return true
    }

    fun remove(feedId: String) {
        sessions.remove(feedId)
        placementSequences.remove(feedId)
    }

    fun clear() {
        sessions.clear()
        placementSequences.clear()
    }

    private companion object {
        const val MAX_FEEDS = 2
    }
}
